### ISSUE TITLE
Issue with streaming decryption for larger files

### ISSUE THREAD

**Issue by kamesh95** [openpgpjs/openpgpjs#974]
I am using openpgpjs for encrypting large files of around 1-2 GB and even more. This works pretty good but when I try to get back the original file by decrypting, it just fails with: **Error: Error decrypting message: Session key decryption failed.** I have tried using fromBinary method with the readable stream of the encrypted file because with readArmored and read methods, it throws **Error: Parameter [message] needs to be of type Message**. I think these two methods do not support stream. Can you maybe give me a proper example of decryption with larger files using stream? I have added my code below. Thanks in advance :)

```
const openpgp = require('openpgp');
const fs = require('fs');
(async () => {
    let options;

    const pubkey = `publickey`;

    const privkey = `privatekey`;

    const passphrase = `phrase`;

    const privKeyObj = (await openpgp.key.readArmored(privkey)).keys[0];
    await privKeyObj.decrypt(passphrase);

    const readableStream = fs.createReadStream('cipher.txt')
    const writableStream = fs.createWriteStream('plain.txt');
    options = {
        message: openpgp.message.fromBinary(readableStream),
        publicKeys: (await openpgp.key.readArmored(pubkey)).keys,
        privateKeys: [privKeyObj],
	format: 'binary'
    };

    const decrypted = await openpgp.decrypt(options);
    const plainText = decrypted.data;
    plainText.pipe(writableStream);

})();

```

**Comment by kamesh95** [comment]
@twiss No that doesn't work because readArmored is not accepting stream args. It throws error that Parameter [message] needs to be of type Message.

**Comment by twiss** [comment]
Could you post your whole code again where you tried that? Because they should accept streams.

**Comment by kamesh95** [comment]
@twiss The code is almost the same just that I am using readArmored method but it throws error
```
const openpgp = require('openpgp');
const fs = require('fs');
(async () => {
    let options;

    const pubkey = `publicKey`; 
    const privkey = `Privatekey`;
    const passphrase = `Kamesh`;

    const privKeyObj = (await openpgp.key.readArmored(privkey)).keys[0];
    await privKeyObj.decrypt(passphrase);

    const readableStream = fs.readFileSync('cipher.txt')
    const writableStream = fs.createWriteStream('plain.txt');
    options = {
        message: openpgp.message.readArmored(readableStream), 
        publicKeys: (await openpgp.key.readArmored(pubkey)).keys, 
        privateKeys: [privKeyObj],
    };

    const encrypted = await openpgp.decrypt(options); // ---> Blows up here
    const ciphertext = encrypted.data; 
    ciphertext.pipe(writableStream);
})();
```

**Comment by kamesh95** [comment]
Also as you suggested using it like: 

`message: await openpgp.message.readArmored(readableStream),`

The above also doesn't work, anyways it would process the entire stream and store it in memory if we used await and waited for it to resolve. I just don't need that, I want to keep the flow async using streams so that the memory consumption is low.

**Comment by twiss** [comment]
> anyways it would process the entire stream and store it in memory if we used await and waited for it to resolve. I just don't need that, I want to keep the flow async using streams so that the memory consumption is low.

That's not necessarily true. I don't mean to be rude, but please stop making assumptions and telling me why it doesn't work and instead start asking why it doesn't work :)

See https://github.com/openpgpjs/openpgpjs/releases/tag/v4.0.0. It will tell you how to fix that.

> The above also doesn't work

I can't help you if you don't tell me what error you get or in what way it doesn't work.

**Comment by kamesh95** [comment]
Sorry my bad. I had a problem with the passphrase as well that I was passing. Resolved it and used await the way you suggested. It does work like a charm. Thanks a lot. You saved me some time ;)

**Comment by twiss** [comment]
No problem, glad it worked!

**Comment by kamesh95** [comment]
Hi @twiss, I am using the decryption method for large files with options `openpgp.config.aead_protect = true` as well as `openpgp.config.allow_unauthenticated_stream = true` But still the memory utilization is very high. I am decrypting a file of size 7 GB and the node process goes as high as my system can handle. It occupies around 90% of the system RAM and has a utilization directly linked with the file size. So in case when I decrypt a 7GB file, it takes 5-6 GB of my RAM and finally crashes. Can you please help me out here?

**Comment by kamesh95** [comment]
```
const openpgp = require('openpgp');
openpgp.config.allow_unauthenticated_stream = true;

const fs = require('fs');
(async () => {
    let options;

    const pubkey = `publickey`;

    const privkey = `privatekey`;

    const passphrase = `phrase`;

    const privKeyObj = (await openpgp.key.readArmored(privkey)).keys[0];
    await privKeyObj.decrypt(passphrase);

    const readableStream = fs.createReadStream('cipher.txt')
    const writableStream = fs.createWriteStream('plain.txt');
    options = {
        message: await openpgp.message.fromBinary(readableStream),
        publicKeys: (await openpgp.key.readArmored(pubkey)).keys,
        privateKeys: [privKeyObj],
	format: 'binary'
    };

    const decrypted = await openpgp.decrypt(options);
    const plainText = decrypted.data;
    plainText.pipe(writableStream);

})();```

**Comment by twiss** [comment]
Hey @kamesh95 :wave: I don't immediately see anything wrong with what you're doing. I think that should work, and so there might be a bug. Could you verify that data is being written to the output file? Also, could you perhaps attach a Chrome debugger and use the Memory panel to take a Heap snapshot, and perhaps Save it and post it here?

**Comment by kamesh95** [comment]
@twiss The decrypted file does grow in size so there's definitely data being written to the output file. Though I already took a snapshot and debugged a bit. I think something is going wrong in the read function for packets. fn at line 37843. Here is the heap snapshot:- [https://a.uguu.se/LNqLxWAelgYL_1571664183620.heapsnapshot](url)

**Comment by kamesh95** [comment]
![memory](https://user-images.githubusercontent.com/16370987/67209358-d6290500-f434-11e9-9d63-822bd5fabdea.png)

**Comment by kamesh95** [comment]
Here's one more heap snapshot I generated today. I have been trying too hard but unfortunately cannot get anything solid on the memory leak. The former heapdump pointed to the `ReadableStreamDefaultController` class and _queue array but when I placed a debug point there, it seems to be working as expected because the queue becomes empty when data is read from it so sometimes _queue has data and sometimes not. Later on I took this another debug trace and now it shows a lot of places where memory is retained - https://drive.google.com/file/d/1KQDcSGBE7ESxBBD56eqyjwZnO2Dsow5g/view?usp=sharing
